# Analysis of Donors dataset

### Loading Dataset

In [1]:
import pandas as pd

file = './donors_data.csv'

df = pd.read_csv(file)
df.head()

Row Id  Row Id.  zipconvert_2  zipconvert_3  zipconvert_4  zipconvert_5  \
0       1       17             0             1             0             0   
1       2       25             1             0             0             0   
2       3       29             0             0             0             1   
3       4       38             0             0             0             1   
4       5       40             0             1             0             0   

   homeowner dummy  NUMCHLD  INCOME  gender dummy  ...  IC15  NUMPROM  \
0                1        1       5             1  ...     1       74   
1                1        1       1             0  ...     4       46   
2                0        2       5             1  ...    13       32   
3                1        1       3             0  ...     4       94   
4                1        1       4             0  ...     7       20   

   RAMNTALL  MAXRAMNT  LASTGIFT  totalmonths  TIMELAG   AVGGIFT  TARGET_B  \
0     102.0       6.0       5.0           29        3  4.857143         1   
1      94.0      12.0      12.0           34        6  9.400000         1   
2      30.0      10.0       5.0           29        7  4.285714         1   
3     177.0      10.0       8.0           30        3  7.080000         0   
4      23.0      11.0      11.0           30        6  7.666667         0   

   TARGET_D  
0       5.0  
1      10.0  
2       5.0  
3       0.0  
4       0.0  

[5 rows x 24 columns]

In [2]:
df.shape

(3120, 24)

In [3]:
df.dropna()

Row Id  Row Id.  zipconvert_2  zipconvert_3  zipconvert_4  zipconvert_5  \
0          1       17             0             1             0             0   
1          2       25             1             0             0             0   
2          3       29             0             0             0             1   
3          4       38             0             0             0             1   
4          5       40             0             1             0             0   
...      ...      ...           ...           ...           ...           ...   
3115    3116    23256             0             0             0             1   
3116    3117    23258             0             1             0             0   
3117    3118    23261             0             0             0             1   
3118    3119    23265             0             0             0             1   
3119    3120    23293             0             0             1             0   

      homeowner dummy  NUMCHLD  INCOME  gender dummy  ...  IC15  NUMPROM  \
0                   1        1       5             1  ...     1       74   
1                   1        1       1             0  ...     4       46   
2                   0        2       5             1  ...    13       32   
3                   1        1       3             0  ...     4       94   
4                   1        1       4             0  ...     7       20   
...               ...      ...     ...           ...  ...   ...      ...   
3115                1        1       4             0  ...    11       23   
3116                0        1       2             0  ...     0       29   
3117                1        1       3             0  ...     6       79   
3118                1        1       7             1  ...     6       27   
3119                1        1       5             1  ...    21       46   

      RAMNTALL  MAXRAMNT  LASTGIFT  totalmonths  TIMELAG    AVGGIFT  TARGET_B  \
0        102.0       6.0       5.0           29        3   4.857143         1   
1         94.0      12.0      12.0           34        6   9.400000         1   
2         30.0      10.0       5.0           29        7   4.285714         1   
3        177.0      10.0       8.0           30        3   7.080000         0   
4         23.0      11.0      11.0           30        6   7.666667         0   
...        ...       ...       ...          ...      ...        ...       ...   
3115      30.0      16.0       4.0           29        5   7.500000         0   
3116      53.0      30.0      30.0           37        9  26.500000         1   
3117     212.0      20.0      20.0           36        3  11.157895         0   
3118      60.0      20.0      20.0           31        9  20.000000         0   
3119     111.0      22.0       9.0           36        5   9.250000         0   

      TARGET_D  
0          5.0  
1         10.0  
2          5.0  
3          0.0  
4          0.0  
...        ...  
3115       0.0  
3116      40.0  
3117       0.0  
3118       0.0  
3119       0.0  

[3120 rows x 24 columns]

### Exploring & Cleaning Data
First, we examine each of the columns in the dataset to see what we can learn about them.

In [4]:
print(list(df.columns))

['Row Id', 'Row Id.', 'zipconvert_2', 'zipconvert_3', 'zipconvert_4', 'zipconvert_5', 'homeowner dummy', 'NUMCHLD', 'INCOME', 'gender dummy', 'WEALTH', 'HV', 'Icmed', 'Icavg', 'IC15', 'NUMPROM', 'RAMNTALL', 'MAXRAMNT', 'LASTGIFT', 'totalmonths', 'TIMELAG', 'AVGGIFT', 'TARGET_B', 'TARGET_D']


#### 'Row Id' and 'Row Id.' 
These appear to be column headers as row indexes from other
sources and don't provide much value by themselves. We will omit these
from our analysis.

#### The zipconvert columns 
These appear to be one-hot-encoded values to show which of the 5 zip codes
a donor lives.

To confirm this, there should be instances where all 4 zip codes are set to zero, indicating
that the donor lives in zipconvert_1. Let's check this with the following code:

In [5]:
df.loc[(df['zipconvert_2'] == 0) & (df['zipconvert_3'] == 0) & (df['zipconvert_4'] == 0) & (df['zipconvert_5'] == 0)]

Row Id  Row Id.  zipconvert_2  zipconvert_3  zipconvert_4  zipconvert_5  \
487      488     3625             0             0             0             0   
1916    1917    14261             0             0             0             0   
2330    2331    17378             0             0             0             0   
2545    2546    18897             0             0             0             0   

      homeowner dummy  NUMCHLD  INCOME  gender dummy  ...  IC15  NUMPROM  \
487                 0        1       1             1  ...     8       50   
1916                0        1       2             0  ...     5       24   
2330                1        1       4             0  ...    17       47   
2545                1        1       4             0  ...    10       44   

      RAMNTALL  MAXRAMNT  LASTGIFT  totalmonths  TIMELAG   AVGGIFT  TARGET_B  \
487       63.0      10.0      10.0           31       12  6.300000         0   
1916      36.0      11.0      10.0           29        7  9.000000         0   
2330      86.0      15.0      13.0           29       12  9.555556         0   
2545     104.0      11.0      11.0           29       10  8.666667         0   

      TARGET_D  
487        0.0  
1916       0.0  
2330       0.0  
2545       0.0  

[4 rows x 24 columns]

#### 'gender dummy' and 'homeowner dummy' are binary values. 

Assumption for homeowner: 1 = homeowner, 0 = non-homeowner.

Assumption for gender: 1 = man, 0 = woman. I assumed this due to the fact that there are more homeowners
who are men compared to women, as seen in the cell below:

In [6]:
df.groupby(['homeowner dummy','gender dummy']).size()

homeowner dummy  gender dummy
0                0                282
                 1                435
1                0                937
                 1               1466
dtype: int64

#### Wealth
The values from this column are 0-9 and appear to be brackets of wealth,
i.e., a net worth of \\$0-\\$50k, \\$51k-\\$80k. However, there is no information
to indicate what these wealth brackets are. 

Assumption: The higher the value for 'wealth', the higher the net worth of the donor.

In [7]:
df[['WEALTH']].value_counts()

WEALTH
8         1700
9          189
5          186
7          180
3          162
6          162
4          153
1          138
2          138
0          112
dtype: int64

#### Income
Similar to 'wealth', the 'income' column appears to represent brackets of income for the donors.
The values range from 1-7 but there is nothing to indicate what values these income brackets equal.

Assumption: The higher the 'income' bracket, the higher the income of the donor.

In [8]:
df[['INCOME']].value_counts()

INCOME
4         1053
5          535
2          468
3          296
1          282
6          246
7          240
dtype: int64

#### Home value
The home value column appears to be the monthly mortgage/rent of a home. This is based on the relatively low values
in this column. The highest value is 5945, which is too low to be the total value of a home.

In [9]:
df['HV'].sort_values(ascending=False)

587     5945
827     5932
2776    5926
800     5908
2818    5888
        ... 
2618       0
39         0
2179       0
1467       0
1664       0
Name: HV, Length: 3120, dtype: int64

There were a few home values of zero, which may indicate that the mortgage was paid off, but
for the purposes of our analysis, we treat these zeros as outliers and thus drop these rows.

In [10]:
df = df[df.HV != 0]
df.shape

(3097, 24)

#### Number of children
This column is self-explanatory. The vast majority of donors had only 1 child, and only 1 had 5 children. 
No donors had zero children.

In [11]:
df[['NUMCHLD']].value_counts()

NUMCHLD
1          2950
2            98
3            31
4            17
5             1
dtype: int64

#### Average gift
This column is the average of gifts per donor. Assumption: This number represents the average number of gifts per donor.

In [12]:
df[['AVGGIFT']].value_counts()

AVGGIFT   
15.000000     96
10.000000     72
20.000000     59
25.000000     44
9.000000      40
              ..
7.275862       1
7.291667       1
7.312500       1
7.362143       1
122.166667     1
Length: 1289, dtype: int64

Our analysis will only require the following columns, so we will exclude the rest.

In [13]:
df = df[['AVGGIFT','INCOME','gender dummy','NUMCHLD','WEALTH','homeowner dummy','HV']]
df

AVGGIFT  INCOME  gender dummy  NUMCHLD  WEALTH  homeowner dummy    HV
0      4.857143       5             1        1       9                1  1399
1      9.400000       1             0        1       7                1   698
2      4.285714       5             1        2       8                0   828
3      7.080000       3             0        1       4                1  1471
4      7.666667       4             0        1       8                1   547
...         ...     ...           ...      ...     ...              ...   ...
3115   7.500000       4             0        1       8                1   697
3116  26.500000       2             0        1       8                0   590
3117  11.157895       3             0        1       7                1  3129
3118  20.000000       7             1        1       8                1  1345
3119   9.250000       5             1        1       4                1   882

[3097 rows x 7 columns]

Here are some statistics on the chosen columns, to include the mean, standard deviation, min and max, and quartiles.

In [14]:
df.describe()

AVGGIFT       INCOME  gender dummy      NUMCHLD       WEALTH  \
count  3097.000000  3097.000000   3097.000000  3097.000000  3097.000000   
mean     10.694372     3.894091      0.610591     1.069422     6.405554   
std       7.452061     1.638844      0.487695     0.348527     2.538122   
min       2.138889     1.000000      0.000000     1.000000     0.000000   
25%       6.368421     3.000000      0.000000     1.000000     5.000000   
50%       9.062500     4.000000      1.000000     1.000000     8.000000   
75%      12.800000     5.000000      1.000000     1.000000     8.000000   
max     122.166667     7.000000      1.000000     5.000000     9.000000   

       homeowner dummy           HV  
count      3097.000000  3097.000000  
mean          0.769454  1149.838231  
std           0.421250   945.007453  
min           0.000000   163.000000  
25%           1.000000   562.000000  
50%           1.000000   826.000000  
75%           1.000000  1343.000000  
max           1.000000  5945.000000

# Question 1:
### What is the gender and number of children of homeowners with home values in the top (75%) quartile?

Unit(s) of analysis: gender, number of children, homeowner, and home value.

Comparison: compute the mode for gender and average for number of children for homeowners in the 75% quartile.

Output: mode for gender, average for number of children in the stated subset.

In [15]:
# first select home values in the top 75% percentile
top_hv = df[df.HV >= df.HV.quantile(.75)]

# now filter on only homeowners (homeowners = 1)
home_top_hv = top_hv.loc[(top_hv['homeowner dummy'] == 1)]

# now get the mode of gender for the filtered dataframe
print('The mode for gender (man=1, woman=0):')
print(home_top_hv['gender dummy'].mode().iat[0])

# now get the mean of number of children for the filtered dataframe
print('\nAverage number of children:')
print(home_top_hv['NUMCHLD'].mean())

# stats for home value column
print('\nStatistics for the home value column:')
print(home_top_hv['HV'].describe())

The mode for gender (man=1, woman=0):
1

Average number of children:
1.0692307692307692

Statistics for the home value column:
count     650.000000
mean     2466.061538
std      1082.179516
min      1343.000000
25%      1663.000000
50%      2121.000000
75%      2903.000000
max      5945.000000
Name: HV, dtype: float64


# Question 2
### How do wealth and income affect average number of gifts?
Unit(s) of analysis: wealth, income, and average number of gifts.

Comparison: find the average for wealth and income in the bottom (25%) quantile and top (75%) quantile and their average
number of gifts in these quantiles.

Output: average number of gifts for wealth and income in the bottom and top quantiles.

In [16]:
# the possible values for income are 1-7. Get the values in the top quantile.
top_incomes = df[df.INCOME >= df.INCOME.quantile(.75)]
top_incomes = top_incomes.INCOME.unique()
top_incomes

array([5, 6, 7], dtype=int64)

In [17]:
# the possible values for income are 1-7. Get the values in the bottom quantile.
bottom_incomes = df[df.INCOME <= df.INCOME.quantile(.25)]
bottom_incomes = bottom_incomes.INCOME.unique()
bottom_incomes

array([1, 3, 2], dtype=int64)

In [18]:
# the possible values for wealth are 0-9. Get the values in the top quantile.
top_wealth = df[df.WEALTH >= df.WEALTH.quantile(.75)]
top_wealth = top_wealth.WEALTH.unique()
top_wealth

array([9, 8], dtype=int64)

In [19]:
# the possible values for wealth are 0-9. Get the values in the bottom quantile.
bottom_wealth = df[df.WEALTH <= df.WEALTH.quantile(.25)]
bottom_wealth = bottom_wealth.WEALTH.unique()
bottom_wealth

array([4, 5, 2, 0, 1, 3], dtype=int64)

In [20]:
# now filter only on top income and top wealth
top_moneys = df.loc[(df['INCOME'].isin(top_incomes)) & (df['WEALTH'].isin(top_wealth))]

# now filter only on bottom income and bottom wealth
bottom_moneys = df.loc[(df['INCOME'].isin(bottom_incomes)) & (df['WEALTH'].isin(bottom_wealth))]

# now get the mean of average number of gifts for the top moneys
print('Average number of gifts among richest donors (75% quantile):')
print(top_moneys['AVGGIFT'].mean())

# now get the mean of average number of gifts for the bottom moneys
print('\nAverage number of gifts among poorest donors (25% quantile):')
print(bottom_moneys['AVGGIFT'].mean())

Average number of gifts among richest donors (75% quantile):
12.39993268808217

Average number of gifts among poorest donors (25% quantile):
9.721989189392765


# Final conclusion
Based on the analysis, the top home values belong to men with an average of 1.1 children. This suggests a negative correlation
between number of children and home values.

Additionally, the richest donors on average give a higher number of gifts. This suggests a positive correlation
between wealth (wealth and income) and number of gifts donated.